kimchi/src/circuits/polynomials/range_check/gate.rs

```diff
@@ -352,49 +586,330 @@ mod tests {
-            biguint_from_hex_le("bca91cf9df6cfd8bd225fd3f46ba2f3f33809d0ee2e7ad338448b4ece7b4f622"),
+        let mut witness = range_check::create_witness::<PallasField>(
+            PallasField::from_hex(
+                "22f6b4e7ecb4488433ade7000000000000000000000000000000000000000000",
+            )
+            .unwrap(),
+            PallasField::from_hex(
+                "e20e9d80333f2fba463ffd000000000000000000000000000000000000000000",
+            )
+            .unwrap(),
+            PallasField::from_hex(
+                "25d28bfd6cdff91ca9bc00000000000000000000000000000000000000000000",
+            )
+            .unwrap(),
         );
 
-        // Invalidate witness
+        // Invalidate witness copy constraint
         witness[5][0] += PallasField::one();
 
         // gates[0] is RangeCheck0
         assert_eq!(
             cs.gates[0].verify_range_check(0, &witness, &cs),
-            Err(String::from(
-                "Invalid RangeCheck0 constraint - permutation failed"
-            ))
+            Err(GateError::InvalidCopyConstraint(GateType::RangeCheck0))
         );
 
-        let mut witness: [Vec<PallasField>; 15] = range_check::create_witness(biguint_from_hex_le(
-            "301a091e9f74cd459a448c311ae47fe2f4311db61ae1cbd2afee0171e2b5ca22",
-        ));
+        // Invalidate witness copy constraint
+        witness[6][1] += PallasField::one();
+
+        // gates[1] is RangeCheck0
+        assert_eq!(
+            cs.gates[1].verify_range_check(1, &witness, &cs),
+            Err(GateError::InvalidCopyConstraint(GateType::RangeCheck0))
+        );
+
+        let mut witness = range_check::create_witness::<PallasField>(
+            PallasField::from_hex(
+                "22cab5e27101eeafd2cbe1000000000000000000000000000000000000000000",
+            )
+            .unwrap(),
+            PallasField::from_hex(
+                "1ab61d31f4e27fe41a318c000000000000000000000000000000000000000000",
+            )
+            .unwrap(),
+            PallasField::from_hex(
+                "449a45cd749f1e091a3000000000000000000000000000000000000000000000",
+            )
+            .unwrap(),
+        );
 
         // Invalidate witness
         witness[8][0] = witness[0][0] + PallasField::one();
 
         // gates[0] is RangeCheck0
         assert_eq!(
             cs.gates[0].verify_range_check(0, &witness, &cs),
-            Err(String::from("Invalid RangeCheck0 constraint"))
+            Err(GateError::InvalidConstraint(GateType::RangeCheck0))
+        );
+
+        // Invalidate witness
+        witness[8][1] = witness[0][1] + PallasField::one();
+
+        // gates[1] is RangeCheck0
+        assert_eq!(
+            cs.gates[1].verify_range_check(1, &witness, &cs),
+            Err(GateError::InvalidConstraint(GateType::RangeCheck0))
+        );
+    }
+
+    #[test]
+    fn verify_range_check0_valid_v0_in_range() {
+        let cs = create_test_constraint_system();
```

Comment:
There's a lot of boilerplate here. It feels like this could be a function like
```rust
enum TestWhich = RangeCheck0 | RangeCheck1 | RangeCheck2

fn test_one(which: TestWhich) -> Result<(), GateError> {
    let cs = create_test_constraint_system();
    let witness = range_check::create_witness::<PallasField>(
        if let RangeCheck0 = which { field } else { PallasField::zero() },
        if let RangeCheck1 = which { field } else { PallasField::zero() },
        if let RangeCheck2 = which { field } else { PallasField::zero() },
    );
    cs.gates[0].verify_range_check(0, &witness, &cs)
}
```

Then your test cases become
```rust
use TestWhich::*;
let field = |i: u64| PallasField::from(i);

assert_eq!(test_one(RangeCheck0, field(2).pow([88]) - f(1)), Ok(()));
assert_eq!(test_one(RangeCheck0, field(2).pow([64]))), Ok(()));
assert_eq!(test_one(RangeCheck0, field(42))), Ok(()));
assert_eq!(test_one(RangeCheck0, field(1))), Ok(()));

assert_eq!(test_one(RangeCheck1, field(2).pow([88]) - f(1)), Ok(()));
assert_eq!(test_one(RangeCheck1, field(2).pow([64]))), Ok(()));
assert_eq!(test_one(RangeCheck1, field(42))), Ok(()));
assert_eq!(test_one(RangeCheck1, field(1))), Ok(()));

assert_eq!(test_one(RangeCheck2, field(2).pow([88]) - f(1)), Ok(()));
assert_eq!(test_one(RangeCheck2, field(2).pow([64]))), Ok(()));
assert_eq!(test_one(RangeCheck2, field(42))), Ok(()));
assert_eq!(test_one(RangeCheck2, field(1))), Ok(()));
```